server/src/main/java/org/elasticsearch/search/internal/ContextIndexSearcher.java

```diff
@@ -192,6 +197,72 @@ public Weight createWeight(Query query, ScoreMode scoreMode, float boost) throws
         }
     }
 
+    @Override
+    protected LeafSlice[] slices(List<LeafReaderContext> leaves) {
+        return computeSlices(leaves, queueSizeBasedExecutor.threadPoolExecutor.getPoolSize(), MINIMUM_DOCS_PER_SLICE);
+    }
+
+    /**
+     * Each computed slice contains at least 10% of the total data in the leaves with a
+     * minimum given by the <code>minDocsPerSlice</code> parameter and the final number
+     * of {@link LeafSlice} will be equal or lower than the max number of slices.
+     */
+    /* pkg private for testing */
+    static LeafSlice[] computeSlices(List<LeafReaderContext> leaves, int maxSliceNum, int minDocsPerSlice) {
+        if (maxSliceNum < 1) {
+            throw new IllegalArgumentException("maxSliceNum must be >= 1 (got " + maxSliceNum + ")");
+        }
+        // total number of documents to be searched
+        final int numDocs = leaves.stream().mapToInt(l -> l.reader().maxDoc()).sum();
+        // percentage of documents per slice, minumum 10%
+        final double percentageDocsPerThread = Math.max(MINIMUM_DOCS_PERCENT_PER_SLICE, 1.0 / maxSliceNum);
+        // compute slices
+        return computeSlices(leaves, Math.max(minDocsPerSlice, (int) (percentageDocsPerThread * numDocs)));
+    }
+
+    private static LeafSlice[] computeSlices(List<LeafReaderContext> leaves, int minDocsPerSlice) {
+        // Make a copy so we can sort:
+        List<LeafReaderContext> sortedLeaves = new ArrayList<>(leaves);
+
+        // Sort by maxDoc, descending:
+        Collections.sort(sortedLeaves, Collections.reverseOrder(Comparator.comparingInt(l -> l.reader().maxDoc())));
```

Comment:
can you remind me why it matters that we sort on number of documents in each segment? we want to avoid having big segments together with very small segments in the same slice?